### PR TITLE
Resource store sql backend watch

### PR DESCRIPTION
**Pull request by chaudyg**
Add watch working across multiple RVs.

### CODE REVIEW

File: pkg/storage/unified/sql/backend.go [hunk 1]
```diff
@@ -500,8 +502,35 @@ func (b *backend) WatchWriteEvents(ctx context.Context) (<-chan *resource.Writte
 	if err := b.Init(); err != nil {
 		return nil, err
 	}
-	// Fetch the lastest RV
-	since, err := fetchLatestRV(ctx, b.sqlDB)
+	// Get the lastest RV
```
Comment (DanCech) [inline pkg/storage/unified/sql/backend.go:505]:
```suggestion
	// Get the latest RV
```

File: pkg/storage/unified/sql/backend.go [hunk 2]
```diff
@@ -500,8 +502,35 @@ func (b *backend) WatchWriteEvents(ctx context.Context) (<-chan *resource.Writte
...
+	if err != nil {
+		return nil, fmt.Errorf("execute SQL template to get the latest resource version: %w", err)
+	}
+	rows, err := b.sqlDB.QueryContext(ctx, query, reqRVs.GetArgs()...)
+	if err != nil {
+		return nil, fmt.Errorf("fetching recent resource versions: %w", err)
+	}
+	defer func() { _ = rows.Close() }()
+
+	for i := 1; rows.Next(); i++ {
```
Comment (DanCech) [inline pkg/storage/unified/sql/backend.go:521]: doesn't look like `i` is actually used for anything

File: pkg/storage/unified/sql/backend.go [hunk 3]
```diff
@@ -527,60 +556,89 @@ func (b *backend) poller(ctx context.Context, since int64, stream chan<- *resour
...
+				}
+				if _, ok := since[reqRVs.Group][reqRVs.Resource]; !ok {
+					since[reqRVs.Group][reqRVs.Resource] = 0 // start from 0 to ensure we do not miss any events
+				}
+			}
 		}
 	}
 }
 
 // fetchLatestRV returns the current maxium RV in the resource table
```
Comment (DanCech) [inline pkg/storage/unified/sql/backend.go:593]:
```suggestion
// fetchLatestRV returns the current maximum RV in the resource table
```

File: pkg/storage/unified/sql/data/resource_history_poll.sql [hunk 4]
```diff
@@ -8,5 +8,14 @@ SELECT
     {{ .Ident "action" | .Into .Response.Action }}
 
     FROM {{ .Ident "resource_history" }}
-    WHERE {{ .Ident "resource_version" }} > {{ .Arg .SinceResourceVersion }}
+    WHERE 1 = 0
+    {{- range $g, $items := .Since }}
+    {{- range $r, $rv := $items }}
+        OR 
+        ( 
+            "group" = "{{ $g }}" AND "resource" = "{{ $r }}" AND "resource_version" > {{ $rv }} AND 1 = 1
```
Comment (DanCech) [inline pkg/storage/unified/sql/data/resource_history_poll.sql:16]: what's the reason for the ` AND 1=1` here?
```suggestion
            "group" = "{{ $g }}" AND "resource" = "{{ $r }}" AND "resource_version" > {{ $rv }}
```

File: pkg/storage/unified/sql/data/resource_history_poll.sql [hunk 5]
```diff
@@ -8,5 +8,14 @@ SELECT
...
+        )
+    {{- end }}
+    {{- end }}
```
Comment (DanCech) [inline pkg/storage/unified/sql/data/resource_history_poll.sql:19]: This is a neat idea, but I'm not sure how well the database will be able to optimize it

Comment (chaudyg) [reply pkg/storage/unified/sql/data/resource_history_poll.sql:19]: Not sure either. The alternative is to poll each resource type independently.

Comment (DanCech) [reply pkg/storage/unified/sql/data/resource_history_poll.sql:19]: Yup, if this drops back to a full table scan we're likely to be better off with per-kind pollers that are able to use the index (and could potentially be started on-demand if we end up with kinds that no-one watches).

**Review by DanCech — APPROVED** [review]
looks good, eventually we may want to do something other than looping through the groupversions synchronously but this will work just fine for now

**Pull request merged**